File: clients/android/app/src/main/kotlin/my/ssdid/drive/domain/repository/OidcRepository.kt
package my.ssdid.drive.domain.repository

import my.ssdid.drive.domain.model.AuthProvider
import my.ssdid.drive.domain.model.User
import my.ssdid.drive.util.Result

interface OidcRepository {
    suspend fun getProviders(tenantSlug: String): Result<List<AuthProvider>>
    suspend fun beginAuthorize(providerId: String): Result<OidcAuthorizeResult>
    suspend fun handleCallback(code: String, state: String): Result<OidcCallbackResult>
    suspend fun completeRegistration(
        providerId: String,
        oidcSub: String,
        email: String,
        name: String,
        keyMaterial: String,
        keySalt: String
    ): Result<User>
}

data class OidcAuthorizeResult(
    val authorizationUrl: String,
    val state: String
)

sealed class OidcCallbackResult {
    data class Authenticated(val user: User) : OidcCallbackResult()
    data class NewUser(val keyMaterial: String, val keySalt: String) : OidcCallbackResult()
}

File: clients/ios/SsdidDrive/SsdidDrive/Domain/Repository/OidcRepository.swift
import Foundation

/// OIDC authorization result
struct OidcAuthorizeResult {
    let authorizationUrl: String
    let state: String
}

/// OIDC callback result
enum OidcCallbackResult {
    case authenticated(User)
    case newUser(keyMaterial: String, keySalt: String)
}

/// Repository for OIDC authentication operations
protocol OidcRepository: AnyObject {
    /// Get available OIDC providers for a tenant
    func getProviders(tenantSlug: String) async throws -> [AuthProvider]

    /// Begin OIDC authorization flow
    func beginAuthorize(providerId: String) async throws -> OidcAuthorizeResult

    /// Handle OIDC callback with authorization code
    func handleCallback(code: String, state: String) async throws -> OidcCallbackResult

    /// Complete registration for new OIDC user
    func completeRegistration(
        keyMaterial: String,
        keySalt: String,
        encryptedMasterKey: String,
        vaultEncryptedMasterKey: String,
        vaultMkNonce: String,
        encryptedPrivateKeys: String,
        publicKeys: [String: String]
    ) async throws -> User
}
